packages/react-reconciler/src/ReactFiberCommitWork.js

```diff
@@ -1141,150 +1141,154 @@ function commitPlacement(finishedWork: Fiber): void {
-          (fundamentalNode: Instance),
-        );
-      } else {
-        removeChild(
-          ((currentParent: any): Instance),
-          (fundamentalNode: Instance),
-        );
-      }
-    } else if (
-      enableSuspenseServerRenderer &&
-      node.tag === DehydratedFragment
-    ) {
-      if (enableSuspenseCallback) {
-        const hydrationCallbacks = finishedRoot.hydrationCallbacks;
-        if (hydrationCallbacks !== null) {
-          const onDeleted = hydrationCallbacks.onDeleted;
-          if (onDeleted) {
-            onDeleted((node.stateNode: SuspenseInstance));
-          }
+function recursivelyUnmountHostNode(
+  node: Fiber,
+  currentParent: Instance,
+  currentParentIsContainer: boolean,
+  finishedRoot: FiberRoot,
+  renderPriorityLevel: ReactPriorityLevel,
+): void {
+  const {stateNode, tag} = node;
+  if (tag === HostComponent || tag === HostText) {
+    commitNestedUnmounts(finishedRoot, node, renderPriorityLevel);
+    // After all the children have unmounted, it is now safe to remove the
+    // node from the tree.
+    if (currentParentIsContainer) {
+      removeChildFromContainer(currentParent, stateNode);
+    } else {
+      removeChild(currentParent, stateNode);
+    }
+    // Don't visit children because we already visited them.
+  } else if (enableFundamentalAPI && tag === FundamentalComponent) {
+    const fundamentalNode = node.stateNode.instance;
+    commitNestedUnmounts(finishedRoot, node, renderPriorityLevel);
+    // After all the children have unmounted, it is now safe to remove the
+    // node from the tree.
+    if (currentParentIsContainer) {
+      removeChildFromContainer(currentParent, fundamentalNode);
+    } else {
+      removeChild(currentParent, fundamentalNode);
+    }
+  } else if (enableSuspenseServerRenderer && tag === DehydratedFragment) {
+    if (enableSuspenseCallback) {
+      const hydrationCallbacks = finishedRoot.hydrationCallbacks;
+      if (hydrationCallbacks !== null) {
+        const onDeleted = hydrationCallbacks.onDeleted;
+        if (onDeleted) {
+          onDeleted(stateNode);
         }
       }
-
-      // Delete the dehydrated suspense boundary and all of its content.
-      if (currentParentIsContainer) {
-        clearSuspenseBoundaryFromContainer(
-          ((currentParent: any): Container),
-          (node.stateNode: SuspenseInstance),
-        );
-      } else {
-        clearSuspenseBoundary(
-          ((currentParent: any): Instance),
-          (node.stateNode: SuspenseInstance),
-        );
-      }
-    } else if (node.tag === HostPortal) {
-      if (node.child !== null) {
-        // When we go into a portal, it becomes the parent to remove from.
-        // We will reassign it back when we pop the portal on the way up.
-        currentParent = node.stateNode.containerInfo;
-        currentParentIsContainer = true;
-        // Visit children because portals might contain host components.
-        node.child.return = node;
```

Comment:
Remove this assignment is probably the riskiest part of the change. Since it's in the commit phase I don't *think* it will break anything but might want to wait to land this until after we've set up the GK in www to quickly revert to an older version of React.

Another option would be to include the assignment in the new, recursive implementation, too, behind its own feature flag.

Let's just make sure whoever does the next sync knows about this.

Reply:
I will track this one, good point though